File: common/src/test/java/org/apache/gravitino/dto/function/TestFunctionDTO.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.gravitino.dto.function;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.google.common.collect.ImmutableMap;
import java.time.Instant;
import java.util.Map;
import org.apache.gravitino.dto.AuditDTO;
import org.apache.gravitino.dto.rel.expressions.LiteralDTO;
import org.apache.gravitino.function.FunctionImpl;
import org.apache.gravitino.function.FunctionType;
import org.apache.gravitino.json.JsonUtils;
import org.apache.gravitino.rel.types.Types;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class TestFunctionDTO {

  @Test
  public void testFunctionDTOSerDe() throws JsonProcessingException {
    AuditDTO audit = AuditDTO.builder().withCreator("user1").withCreateTime(Instant.now()).build();

    // Create function parameter with default value
    FunctionParamDTO param1 =
        FunctionParamDTO.builder()
            .withName("x")
            .withDataType(Types.IntegerType.get())
            .withComment("input parameter")
            .build();

    FunctionParamDTO param2 =
        FunctionParamDTO.builder()
            .withName("y")
            .withDataType(Types.FloatType.get())
            .withDefaultValue(
                LiteralDTO.builder().withDataType(Types.FloatType.get()).withValue("1.0").build())
            .build();

    // Create SQL implementation
    SQLImplDTO sqlImpl =
        new SQLImplDTO(
            FunctionImpl.RuntimeType.SPARK.name(),
            null,
            ImmutableMap.of("key", "value"),
            "SELECT x + y");

    // Create function definition
    FunctionDefinitionDTO definition =
        FunctionDefinitionDTO.builder()
            .withParameters(new FunctionParamDTO[] {param1, param2})
            .withImpls(new FunctionImplDTO[] {sqlImpl})
            .build();

    // Create scalar function DTO
    FunctionDTO scalarFunction =
        FunctionDTO.builder()
            .withName("add_func")
            .withFunctionType(FunctionType.SCALAR)
            .withDeterministic(true)
            .withComment("A simple add function")
            .withReturnType(Types.FloatType.get())
            .withDefinitions(new FunctionDefinitionDTO[] {definition})
            .withAudit(audit)
            .build();

    // Serialize and deserialize
    String json = JsonUtils.objectMapper().writeValueAsString(scalarFunction);
    FunctionDTO deserialized = JsonUtils.objectMapper().readValue(json, FunctionDTO.class);

    Assertions.assertEquals(scalarFunction.name(), deserialized.name());
    Assertions.assertEquals(scalarFunction.functionType(), deserialized.functionType());
    Assertions.assertEquals(scalarFunction.deterministic(), deserialized.deterministic());
    Assertions.assertEquals(scalarFunction.comment(), deserialized.comment());
    Assertions.assertEquals(scalarFunction.returnType(), deserialized.returnType());
    Assertions.assertEquals(scalarFunction.definitions().length, deserialized.definitions().length);
  }

  @Test
  public void testTableFunctionDTOSerDe() throws JsonProcessingException {
    AuditDTO audit = AuditDTO.builder().withCreator("user1").withCreateTime(Instant.now()).build();

    // Create return columns for table function
    FunctionColumnDTO col1 =
        FunctionColumnDTO.builder()
            .withName("id")
            .withDataType(Types.IntegerType.get())
            .withComment("id column")
            .build();

    FunctionColumnDTO col2 =
        FunctionColumnDTO.builder().withName("name").withDataType(Types.StringType.get()).build();

    // Create Java implementation
    JavaImplDTO javaImpl =
        new JavaImplDTO(
            FunctionImpl.RuntimeType.SPARK.name(),
            FunctionResourcesDTO.builder()
                .withJars(new String[] {"hdfs://path/to/jar.jar"})
                .build(),
            ImmutableMap.of(),
            "com.example.TableFunction");

    // Create function definition
    FunctionDefinitionDTO definition =
        FunctionDefinitionDTO.builder()
            .withParameters(new FunctionParamDTO[] {})
            .withImpls(new FunctionImplDTO[] {javaImpl})
            .build();

    // Create table function DTO
    FunctionDTO tableFunction =
        FunctionDTO.builder()
            .withName("table_func")
            .withFunctionType(FunctionType.TABLE)
            .withDeterministic(false)
            .withComment("A table function")
            .withReturnColumns(new FunctionColumnDTO[] {col1, col2})
            .withDefinitions(new FunctionDefinitionDTO[] {definition})
            .withAudit(audit)
            .build();

    // Serialize and deserialize
    String json = JsonUtils.objectMapper().writeValueAsString(tableFunction);
    FunctionDTO deserialized = JsonUtils.objectMapper().readValue(json, FunctionDTO.class);

    Assertions.assertEquals(tableFunction.name(), deserialized.name());
    Assertions.assertEquals(FunctionType.TABLE, deserialized.functionType());
    Assertions.assertEquals(2, deserialized.returnColumns().length);
  }

  @Test
  public void testFunctionImplDTOSerDe() throws JsonProcessingException {
    Map<String, String> props = ImmutableMap.of("key", "value");
    FunctionResourcesDTO resources =
        FunctionResourcesDTO.builder()
            .withJars(new String[] {"hdfs://path/to/jar.jar"})
            .withFiles(new String[] {"hdfs://path/to/file.txt"})
            .build();

    // Test SQL implementation
    SQLImplDTO sqlImpl = new SQLImplDTO("SPARK", resources, props, "SELECT 1");
    String sqlJson = JsonUtils.objectMapper().writeValueAsString(sqlImpl);
    FunctionImplDTO deserializedSql =
        JsonUtils.objectMapper().readValue(sqlJson, FunctionImplDTO.class);
    Assertions.assertTrue(deserializedSql instanceof SQLImplDTO);
    Assertions.assertEquals(FunctionImpl.Language.SQL, deserializedSql.language());
    Assertions.assertEquals("SELECT 1", ((SQLImplDTO) deserializedSql).getSql());

    // Test Java implementation
    JavaImplDTO javaImpl = new JavaImplDTO("SPARK", resources, props, "com.example.MyUDF");
    String javaJson = JsonUtils.objectMapper().writeValueAsString(javaImpl);
    FunctionImplDTO deserializedJava =
        JsonUtils.objectMapper().readValue(javaJson, FunctionImplDTO.class);
    Assertions.assertTrue(deserializedJava instanceof JavaImplDTO);
    Assertions.assertEquals(FunctionImpl.Language.JAVA, deserializedJava.language());
    Assertions.assertEquals("com.example.MyUDF", ((JavaImplDTO) deserializedJava).getClassName());

    // Test Python implementation
    PythonImplDTO pythonImpl =
        new PythonImplDTO(
            "SPARK", resources, props, "my_handler", "def my_handler(x): return x * 2");
    String pythonJson = JsonUtils.objectMapper().writeValueAsString(pythonImpl);
    FunctionImplDTO deserializedPython =
        JsonUtils.objectMapper().readValue(pythonJson, FunctionImplDTO.class);
    Assertions.assertTrue(deserializedPython instanceof PythonImplDTO);
    Assertions.assertEquals(FunctionImpl.Language.PYTHON, deserializedPython.language());
    Assertions.assertEquals("my_handler", ((PythonImplDTO) deserializedPython).getHandler());
  }

  @Test
  public void testFunctionParamDTOWithDefaultValue() throws JsonProcessingException {
    // Test parameter without default value
    FunctionParamDTO paramWithoutDefault =
        FunctionParamDTO.builder()
            .withName("x")
            .withDataType(Types.IntegerType.get())
            .withComment("no default")
            .build();

    String json1 = JsonUtils.objectMapper().writeValueAsString(paramWithoutDefault);
    FunctionParamDTO deserialized1 =
        JsonUtils.objectMapper().readValue(json1, FunctionParamDTO.class);
    Assertions.assertEquals("x", deserialized1.name());
    Assertions.assertEquals(Types.IntegerType.get(), deserialized1.dataType());

    // Test parameter with default value
    LiteralDTO defaultValue =
        LiteralDTO.builder().withDataType(Types.IntegerType.get()).withValue("10").build();

    FunctionParamDTO paramWithDefault =
        FunctionParamDTO.builder()
            .withName("y")
            .withDataType(Types.IntegerType.get())
            .withDefaultValue(defaultValue)
            .build();

    String json2 = JsonUtils.objectMapper().writeValueAsString(paramWithDefault);
    FunctionParamDTO deserialized2 =
        JsonUtils.objectMapper().readValue(json2, FunctionParamDTO.class);
    Assertions.assertEquals("y", deserialized2.name());
    Assertions.assertNotNull(deserialized2.defaultValue());
  }

  @Test
  public void testFunctionResourcesDTOSerDe() throws JsonProcessingException {
    FunctionResourcesDTO resources =
        FunctionResourcesDTO.builder()
            .withJars(new String[] {"hdfs://path/to/jar1.jar", "hdfs://path/to/jar2.jar"})
            .withFiles(new String[] {"hdfs://path/to/file.txt"})
            .withArchives(new String[] {"hdfs://path/to/archive.zip"})
            .build();

    String json = JsonUtils.objectMapper().writeValueAsString(resources);
    FunctionResourcesDTO deserialized =
        JsonUtils.objectMapper().readValue(json, FunctionResourcesDTO.class);

    Assertions.assertArrayEquals(resources.getJars(), deserialized.getJars());
    Assertions.assertArrayEquals(resources.getFiles(), deserialized.getFiles());
    Assertions.assertArrayEquals(resources.getArchives(), deserialized.getArchives());
  }

  @Test
  public void testFunctionDefinitionDTOSerDe() throws JsonProcessingException {
    FunctionParamDTO param =
        FunctionParamDTO.builder().withName("input").withDataType(Types.StringType.get()).build();

    SQLImplDTO impl = new SQLImplDTO("SPARK", null, null, "SELECT UPPER(input)");

    FunctionDefinitionDTO definition =
        FunctionDefinitionDTO.builder()
            .withParameters(new FunctionParamDTO[] {param})
            .withImpls(new FunctionImplDTO[] {impl})
            .build();

    String json = JsonUtils.objectMapper().writeValueAsString(definition);
    FunctionDefinitionDTO deserialized =
        JsonUtils.objectMapper().readValue(json, FunctionDefinitionDTO.class);

    Assertions.assertEquals(1, deserialized.parameters().length);
    Assertions.assertEquals("input", deserialized.parameters()[0].name());
    Assertions.assertEquals(1, deserialized.impls().length);
  }
}

File: clients/client-python/tests/unittests/dto/test_function_dto.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License a
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Tests for Function DTOs."""

import unittest

from gravitino.api.function.function_type import FunctionType
from gravitino.api.rel.types.types import Types
from gravitino.dto.audit_dto import AuditDTO
from gravitino.dto.function.function_column_dto import FunctionColumnDTO
from gravitino.dto.function.function_definition_dto import FunctionDefinitionDTO
from gravitino.dto.function.function_dto import FunctionDTO
from gravitino.dto.function.function_impl_dto import (
    SQLImplDTO,
    PythonImplDTO,
    JavaImplDTO,
)
from gravitino.dto.function.function_param_dto import FunctionParamDTO
from gravitino.dto.function.function_resources_dto import FunctionResourcesDTO
from gravitino.dto.rel.expressions.literal_dto import LiteralDTO


class TestFunctionDTO(unittest.TestCase):
    """Tests for Function DTOs."""

    def test_function_dto(self):
        """Test FunctionDTO serialization and deserialization."""
        audit = AuditDTO(
            "creator", "2022-01-01T00:00:00Z", "modifier", "2022-01-01T00:00:00Z"
        )
        params = [FunctionParamDTO(_name="param1", _data_type=Types.IntegerType.get())]
        impl = SQLImplDTO(
            _runtime="SPARK",
            _sql="SELECT param1 + 1",
            _resources=None,
            _properties={},
        )
        definition = FunctionDefinitionDTO(
            _parameters=params, _return_type=Types.IntegerType.get(), _impls=[impl]
        )

        function_dto = FunctionDTO(
            _name="func1",
            _definitions=[definition],
            _function_type=FunctionType.SCALAR,
            _deterministic=True,
            _comment="comment",
            _audit=audit,
        )

        json_str = function_dto.to_json()
        deserialized = FunctionDTO.from_json(json_str)
        self.assertEqual(function_dto.name(), deserialized.name())
        self.assertEqual(function_dto.comment(), deserialized.comment())
        self.assertEqual(
            function_dto.audit_info().creator(), deserialized.audit_info().creator()
        )
        self.assertEqual(
            function_dto.definitions()[0].impls()[0].sql(),
            deserialized.definitions()[0].impls()[0].sql(),
        )
        self.assertEqual(
            function_dto.definitions()[0].return_type(),
            deserialized.definitions()[0].return_type(),
        )

    def test_function_definition_dto(self):
        """Test FunctionDefinitionDTO serialization and deserialization."""
        params = [FunctionParamDTO(_name="param1", _data_type=Types.IntegerType.get())]
        impl = SQLImplDTO(
            _runtime="SPARK",
            _sql="SELECT param1 + 1",
            _resources=None,
            _properties={},
        )
        definition = FunctionDefinitionDTO(
            _parameters=params, _return_type=Types.IntegerType.get(), _impls=[impl]
        )

        json_str = definition.to_json()
        deserialized = FunctionDefinitionDTO.from_json(json_str)
        self.assertEqual(
            definition.parameters()[0].name(), deserialized.parameters()[0].name()
        )
        self.assertEqual(definition.impls()[0].sql(), deserialized.impls()[0].sql())
        self.assertEqual(definition.return_type(), deserialized.return_type())

    def test_function_param_dto_default_value(self):
        """Test FunctionParamDTO supports defaultValue serdes."""
        default_value = (
            LiteralDTO.builder()
            .with_data_type(Types.IntegerType.get())
            .with_value("0")
            .build()
        )
        param = FunctionParamDTO(
            _name="x",
            _data_type=Types.IntegerType.get(),
            _comment="comment",
            _default_value=default_value,
        )

        json_str = param.to_json()
        deserialized = FunctionParamDTO.from_json(json_str)
        self.assertEqual(param, deserialized)
        self.assertIsNotNone(deserialized.default_value())

    def test_function_definition_hash_with_impls(self):
        """Test FunctionDefinitionDTO hash works when impl list is not empty."""
        definition = FunctionDefinitionDTO(
            _parameters=[
                FunctionParamDTO(_name="param1", _data_type=Types.IntegerType.get())
            ],
            _return_type=Types.IntegerType.get(),
            _impls=[
                SQLImplDTO(
                    _runtime="SPARK",
                    _sql="SELECT param1 + 1",
                    _resources=None,
                    _properties={},
                )
            ],
        )

        self.assertIsInstance(hash(definition), int)

    def test_function_impl_dto(self):
        """Test FunctionImplDTO serialization and deserialization."""
        sql_impl = SQLImplDTO(
            _runtime="SPARK",
            _sql="SELECT 1",
            _resources=None,
            _properties={},
        )
        json_str = sql_impl.to_json()
        deserialized = SQLImplDTO.from_json(json_str)
        self.assertEqual(sql_impl.sql(), deserialized.sql())
        self.assertEqual(sql_impl.runtime(), deserialized.runtime())

        python_impl = PythonImplDTO(
            _runtime="PYTHON",
            _handler="test_module.test_func",
            _code_block="def test_func(): pass",
            _resources=None,
            _properties={},
        )
        json_str = python_impl.to_json()
        deserialized = PythonImplDTO.from_json(json_str)
        self.assertEqual(python_impl.handler(), deserialized.handler())
        self.assertEqual(python_impl.code_block(), deserialized.code_block())

        java_impl = JavaImplDTO(
            _runtime="JAVA",
            _class_name="com.test.TestClass",
            _resources=None,
            _properties={},
        )
        json_str = java_impl.to_json()
        deserialized = JavaImplDTO.from_json(json_str)
        self.assertEqual(java_impl.class_name(), deserialized.class_name())

    def test_function_resources_dto(self):
        """Test FunctionResourcesDTO serialization and deserialization."""
        resources = FunctionResourcesDTO(_jars=["v1"], _files=None, _archives=None)
        json_str = resources.to_json()
        deserialized = FunctionResourcesDTO.from_json(json_str)
        self.assertEqual(resources.jars(), deserialized.jars())

    def test_function_column_dto(self):
        """Test FunctionColumnDTO serialization and deserialization."""
        column = FunctionColumnDTO(_name="col1", _data_type=Types.IntegerType.get())
        json_str = column.to_json()
        deserialized = FunctionColumnDTO.from_json(json_str)
        self.assertEqual(column.name(), deserialized.name())
        self.assertEqual(column.data_type(), deserialized.data_type())

    def test_function_type_cannot_be_null(self):
        """Test FunctionDTO rejects null functionType during deserialization."""
        json_str = (
            '{"name":"func1","functionType":null,"deterministic":true,"definitions":[]}'
        )
        with self.assertRaises(ValueError):
            FunctionDTO.from_json(json_str)
